### PR TITLE
Native Anderson Acceleration

### PR DESCRIPTION
**Pull request by jpjones6**
Implements a fully native arbitrary depth Anderson acceleration algorithm with efficient (direct) calculation of Anderson coefficients.

### CODE REVIEW

File: src/AndersonAcceleration.f90 [hunk 1]
```diff
@@ -6,55 +6,50 @@
...
+  !> Starting iteration for Anderson
+  INTEGER(SIK) :: start=1
+  !> Value of mixing parameter
   REAL(SRK) :: beta=0.0_SRK
-  !> Pointer to solution vector, x
-  CLASS(VectorType),POINTER :: X=>NULL()
-  !> Timer to measure solution time
-  TYPE(TimerType) :: SolveTime
+  !> Initial iterates
+  REAL(SRK),ALLOCATABLE :: x(:,:)
```
Comment (bscollin) [inline src/AndersonAcceleration.f90:44]: I thought we were changing the solver to use vector classes.

Comment (jpjones6) [reply src/AndersonAcceleration.f90:44]: We had talked about it, and I tried that. Switched to Vector classes and BLAS calls. Made the code slower (only 4% but still). Without an increase in performance I don't see much of a point in using vector classes here, so I switched back. Everything we want to accelerate in MPACT comes as some form/shape of REAL array (not vectors) which needs to be repacked into a 1D array. Any packing/repacking and the Anderson code itself is much cleaner when not using vector classes. Also, if anyone outside of the MPACT team wanted to use the Anderson ability not having to go through vector types/classes may help a beginner use this more easily.

File: src/AndersonAcceleration.f90 [hunk 2]
```diff
@@ -72,183 +67,239 @@ MODULE AndersonAccelerationTypes
     PROCEDURE,PASS :: reset => reset_AndersonAccelerationType
 ENDTYPE AndersonAccelerationType
 
-!> Exception Handler for use in Anderson Acceleration
-TYPE(ExceptionHandlerType),SAVE :: eAndersonAccelerationType
-
 !> Name of module
-CHARACTER(LEN=*),PARAMETER :: modName='ANDERSONACCELERATIONTYPE'
+CHARACTER(LEN=*),PARAMETER :: modName='AndersonAccelerationTypes'
```
Comment (stimpsonsg) [inline src/AndersonAcceleration.f90:71]: why this change?  I'm not exactly opposed to it, but out of Futility, there are only three that are not full caps.

We need them to be consistent (so those other ones will need to change to all caps too) or we change all to camelcase.

Comment (jpjones6) [reply src/AndersonAcceleration.f90:71]: I'm used to seeing camelcase on the MPACT side, so when I added the missing s on the end of the mod name I switched it over. I don't REALLY have a preference, but the logical choice to me would be to be consistent with MPACT.

File: src/AndersonAcceleration.f90 (hunk 1 above)
Comment (stimpsonsg) [reply src/AndersonAcceleration.f90:44]: are any of the arrays more than 2 dimensional?

**Comment by jpjones6** [comment]
New version using vector types and BLAS is up. Caveats: 

1) Only RealVectorType and PetcVectorTypes are usable right now. NativeDistributed and Trilinos type vectors are both missing certain interfaces to needed BLAS routines.

2) Creation of the storage vectors is deferred until the first call of the set/reset Anderson routine. Did this because that is the first time a vector is actually handed to Anderson for any purpose. Could alternatively move it to the init routine with the addition of a parameter list entry specifying what vector type will be used. Didn't like that option since that will require consistency between where Anderson is inited and set/reset in the client code....I don't trust that will always happen plus it seems less flexible.

File: src/AndersonAcceleration.f90 [hunk 3]
```diff
@@ -85,7 +90,7 @@ SUBROUTINE init_AndersonAccelerationType(solver,ce,Params)
   CHARACTER(LEN=*),PARAMETER :: myName='init_AndersonAccelerationType'
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
   TYPE(ParamType),INTENT(IN) :: Params
-  TYPE(FutilityComputingEnvironment),INTENT(IN) :: ce
+  TYPE(FutilityComputingEnvironment),TARGET :: ce
```
Comment (bscollin) [inline src/AndersonAcceleration.f90:93]: Don't we still need intent on this?

File: src/AndersonAcceleration.f90 [hunk 4]
```diff
@@ -205,40 +219,36 @@ SUBROUTINE step_AndersonAccelerationType(solver,x_new)
...
-      ENDDO
+      CALL BLAS_copy(solver%r(2),tmpvec)
+      CALL BLAS_axpy(solver%r(1),tmpvec,-1.0_SRK)
+      tmpA=BLAS_dot(tmpvec,tmpvec)
+      tmpb=BLAS_dot(solver%r(2),tmpvec)
       solver%alpha(1)=tmpb/tmpA
     ELSEIF(depth_s > 1) THEN
       !Construct coefficient matrix, right hand side, and solve for fit coefficients
       DO i=1,depth_s
+        CALL BLAS_copy(solver%r(i),x_new)
```
Comment (bscollin) [inline src/AndersonAcceleration.f90:242]: You made a comment before that the blas routines make the math less readable... could always add a comment to communicate exactly what math you are trying to implement

File: src/AndersonAcceleration.f90 [hunk 5]
```diff
@@ -251,23 +261,33 @@ SUBROUTINE step_AndersonAccelerationType(solver,x_new)
...
     ENDDO
 
-    !Get accelerated solution
-    x_new(:)=0.0_SRK
-    DO i=1,depth_s+1
-      x_new(:)=x_new(:)+solver%alpha(i)*(solver%x(:,i)+solver%beta*solver%r(:,i))
-    ENDDO
+    !Ensure Anderson coefficient solve did not fail due to bad input vector
+    IF(ISNAN(solver%alpha(depth_s+1))) THEN
+      !Bad Anderson coefficient solve, revert to under-relaxation and reset Anderson solver
```
Comment (bscollin) [inline src/AndersonAcceleration.f90:266]: should we make sure we don't drop into some infinite loop where we keep failing?

File: src/AndersonAcceleration.f90 [hunk 6]
```diff
@@ -278,15 +298,63 @@ SUBROUTINE step_AndersonAccelerationType(solver,x_new)
...
+  CLASS(VectorType),INTENT(IN) :: x
+
+  TYPE(ParamType) :: vecparams
+  INTEGER(SIK) :: i,j,m
 
-  INTEGER(SIK) :: i,j
+  REQUIRE(x%n == solver%n)
 
-  REQUIRE(SIZE(x) == solver%n)
+  !If this is the first call to set/reset must actually create vectors of needed type
```
Comment (bscollin) [inline src/AndersonAcceleration.f90:310]: Vector types has a base capability to create a new vector identical to an old vector: `VectorResemble`

Comment (bscollin) [reply src/AndersonAcceleration.f90:310]: also... what blas interface do you need that doesn't exist?

File: src/AndersonAcceleration.f90 (hunk 3 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:93]: Let's put it in. We are missing Intent on the futility computing env in multiple locations in MPACT (was just emulating an init routine in MPACT).

File: src/AndersonAcceleration.f90 (hunk 5 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:266]: We shouldn't end up in an infinite loop..."shouldn't". I've made the behavior here when Anderson fails to just do simple under-relaxation (like default) using the mixing parameter, and then that call to reset will throw away all the past solutions (that would have caused the failure). If we ran into a case where it would fail over and over it would just act like default.

Comment (jpjones6) [reply src/AndersonAcceleration.f90:266]: I need to add a test for this ability in the unit test.

File: src/AndersonAcceleration.f90 (hunk 6 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:310]: Okay, I will look at the resemble routine and see about simplifying the reset/set function. For Trilinos vectors it is dot product (as hard as that is to believe I am looking right at the interface, it's got NativeDistributed, Petsc, and Real...no Trilinos). For NativeDistributed it's the _scal routine. Shouldn't be hard to add.

Comment (jpjones6) [reply src/AndersonAcceleration.f90:310]: On the subject of BLAS interfaces. The vector interfaces to them have their arguments out of order...for instance _axpy for vectors actually has the arguments in the order xya...a little OCD for sure, but annoying.

File: src/AndersonAcceleration.f90 (hunk 5 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:266]: Improved the rescue to look for when Alpha = 0.0 this signals that a duplicate or other bad iterate has been handed in and added testing.

File: src/AndersonAcceleration.f90 (hunk 6 above)
Comment (bscollin) [reply src/AndersonAcceleration.f90:308]: None of this was fixed yet

File: src/AndersonAcceleration.f90 (hunk 4 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:240]: Since the math is simple enough here I don't think we need to go that far.

File: src/AndersonAcceleration.f90 (hunk 1 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90]: Everything is changed over now.

File: src/AndersonAcceleration.f90 (hunk 6 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:308]: Latest push adds resemble routines for allocatable vectors and arrays of vectors as well as associated testing. Also changed Anderson over to using them with a short check for vector types that are not currently fully supported in BLAS.

File: src/AndersonAcceleration.f90 [hunk 7]
```diff
@@ -6,55 +6,54 @@
...
-!>  - Trilinos
-!>
-!> @par EXAMPLES
-!> @code
-!>
-!> @endcode
-!++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++!
+!> @brief A Fortran 2003 module that encompasses the native implementation of
+!>        Anderson Acceleration in Futility.
 MODULE AndersonAccelerationTypes
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:11]: A pretty minor one, but our style guide says modules should be named "<filename without extension>Module"... so for this, "AndersonAccelerationModule".  Granted, virtually nothing in Futility has been updated to follow this, but since this module is unused in downstream code so far, let's try to start it off correctly.

File: src/AndersonAcceleration.f90 [hunk 8]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
     PROCEDURE,PASS :: reset => reset_AndersonAccelerationType
 ENDTYPE AndersonAccelerationType
 
-!> Exception Handler for use in Anderson Acceleration
-TYPE(ExceptionHandlerType),SAVE :: eAndersonAccelerationType
-
 !> Name of module
-CHARACTER(LEN=*),PARAMETER :: modName='ANDERSONACCELERATIONTYPE'
-
+CHARACTER(LEN=*),PARAMETER :: modName='AndersonAccelerationTypes'
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:75]: And you'll want to change this to match the updated module name

File: src/AndersonAcceleration.f90 [hunk 9]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
 !
 !===============================================================================
 CONTAINS
 !
 !-------------------------------------------------------------------------------
-!> @brief Initializes the Anderson Acceleration Type with a parameter list
+!> @brief Initializes the Anderson Acceleration Type with an input parameter list
 !> @param pList the parameter list
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:82]: looks like this is unneeded since you have `Params`

File: src/AndersonAcceleration.f90 [hunk 10]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
-  IF(.NOT. MPIEnv%isInit()) THEN
-    CALL eAndersonAccelerationType%raiseError('Incorrect input to '// &
-        modName//'::'//myName//' - MPI Environment is not initialized!')
-  ELSE
-    solver%MPIparallelEnv => MPIEnv
-  ENDIF
+  REQUIRE(Params%has('AndersonAccelerationType->n'))
+  REQUIRE(Params%has('AndersonAccelerationType->depth'))
+  REQUIRE(Params%has('AndersonAccelerationType->beta'))
+  REQUIRE(Params%has('AndersonAccelerationType->start'))
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:99]: Just out of curiosity, is it possible to set reasonable defaults for some of these?  Or are the values fairly application-specific?

File: src/AndersonAcceleration.f90 [hunk 11]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
-  CALL validParams%get('AndersonAccelerationType->depth',depth)
-  CALL validParams%get('AndersonAccelerationType->beta',beta)
-  CALL validParams%get('AndersonAccelerationType->start',start)
-  CALL validParams%clear()
+  CALL Params%get('AndersonAccelerationType->n',solver%n)
+  CALL Params%get('AndersonAccelerationType->depth',solver%depth)
+  CALL Params%get('AndersonAccelerationType->beta',solver%beta)
+  CALL Params%get('AndersonAccelerationType->start',solver%start)
 
   IF(.NOT. solver%isInit) THEN
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:107]: So, instead of a giant IF block, just `REQUIRE(.NOT.solver%isInit)` instead

File: src/AndersonAcceleration.f90 [hunk 12]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
+      !Construct coefficient matrix, right hand side, and solve for fit coefficients
+      DO i=1,depth_s
+        CALL BLAS_copy(solver%r(i),x_new)
+        CALL BLAS_axpy(solver%r(depth_s+1),x_new,-1.0_SRK)
+        tmpb=BLAS_dot(solver%r(depth_s+1),x_new)
+        CALL solver%LS%b%set(i,tmpb)
+        DO k=1,i
+          CALL BLAS_copy(solver%r(depth_s+1),solver%tmpvec)
+          CALL BLAS_axpy(solver%r(k),solver%tmpvec,-1.0_SRK)
+          tmpA=BLAS_dot(solver%tmpvec,x_new)
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:247]: bugs me that all the BLAS stuff are subroutines except this one... but I know we can't do anything about that

File: src/AndersonAcceleration.f90 [hunk 13]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
-!> @param solver The anderson solver to act on
-!>
-!> This routine resets the counters for the anderson acceleration
+!> @brief Set or reset the initial iterate and linear solver state for the Anderson solver
+!> @param solver The Anderson solver to act on
+!> @param x Initial iterate solve is starting from
 !>
-SUBROUTINE reset_AndersonAccelerationType(solver)
+SUBROUTINE reset_AndersonAccelerationType(solver,x)
+  CHARACTER(LEN=*),PARAMETER :: myName='reset_AndersonAccelerationType'
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:300]: slide this down below the dummy arguments

File: src/AndersonAcceleration.f90 [hunk 14]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
+  REQUIRE(x%n == solver%n)
+
+  !If this is the first call to set/reset must actually create vectors of needed type
+  IF(solver%s == 0) THEN
+    IF(.NOT.ALLOCATED(solver%x)) THEN
+
+      !!!TODO Once missing BLAS interfaces have been added for the following vector types
+      !do away with this error catch to allow use of these with Anderson Acceleration.
+#ifdef HAVE_MPI
+      SELECT TYPE(x);TYPE IS(NativeDistributedVectorType)
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:315]: `TYPE IS(...)` should go on its own line

File: src/AndersonAcceleration.f90 [hunk 15]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
+        CALL solver%ce%exceptHandler%raiseError('Incorrect call to '//modName// &
+            '::'//myName//' - Input vector type not supported!')
 #endif
+#ifdef FUTILITY_HAVE_Trilinos
+      TYPE IS(TrilinosVectorType)
+        CALL solver%ce%exceptHandler%raiseError('Incorrect call to '//modName// &
+            '::'//myName//' - Input vector type not supported!')
+      ENDSELECT
+#endif
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:324]: So the fact that SELECTTYPE and ENDSELECT are in 2 different sets of preproc macros isn't good... I'd recommend this:

```
SELECTTYPE
#ifdef HAVE_MPI
  !Check for native distributed vector
#endif
#ifdef FUTILITY_HAVE_Trilinos
  !Check for trilinos vector
#endif
CLASS DEFAULT
  CONTINUE
ENDSELECT
```

File: src/AndersonAcceleration.f90 [hunk 16]
```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
...
+  ENDIF
+
+  !Reset iteration counter
+  solver%s=0_SIK
+
+  !Grab initial iterate to initiate Anderson from
+  CALL BLAS_copy(x,solver%x(1))
+
+  !Reset Anderson coefficient matrix
+  IF(solver%LS%isinit) THEN
```
Comment (aarograh) [inline src/AndersonAcceleration.f90:343]: Should we just REQUIRE this?  Seems like you shouldn't be calling reset on something that was never initialized to begin with

File: src/AndersonAcceleration.f90 (hunk 7 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:11]: I'm glad you mentioned this. I was leaning toward changing it, but since the module already existed before I gutted and remade it I didn't want to change it without a second opinion.

File: src/AndersonAcceleration.f90 (hunk 8 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:75]: Got it, also changed the copybriefs/details

File: src/AndersonAcceleration.f90 (hunk 9 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:82]: Removed it

File: src/AndersonAcceleration.f90 (hunk 10 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:99]: So, defaults are/will be set in XMLProc for a normal run by a user. Depth=1, beta=0.5, start=1 is a conservative choice that should at least work "okay" for essentially any problem.

Comment (aarograh) [reply src/AndersonAcceleration.f90:99]: Ok.  I'm kind of in favor of setting these defaults internally as opposed to using XMLProc... thoughts @stimpsonsg @bscollin ?

File: src/AndersonAcceleration.f90 (hunk 11 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:107]: I like it.

File: src/AndersonAcceleration.f90 (hunk 13 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:300]: Are we changing our format to do it that way?

File: src/AndersonAcceleration.f90 (hunk 14 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:315]: I agree. I'll change it here. We have a LOT of instances in MPACT with the ; for this.

File: src/AndersonAcceleration.f90 (hunk 13 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:300]: We've pretty much always done it that way...  Most often you see this:

```
SUBROUTINE stuff(arg1,arg2)
  TYPE(*) :: arg1
  TYPE(*) :: arg2
  !
  CHARACTER(LEN=*),PARAMETER :: myName='stuff'

  !Rest of subroutine
ENDSUBROUTINE stuff
```

File: src/AndersonAcceleration.f90 (hunk 15 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:324]: Whoa, yeah that's a bug. That should fail the ci.

File: src/AndersonAcceleration.f90 (hunk 14 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:315]: So the standard allows the `;` if there is only one branch.  But since you have 2 (or 3 with my proposed change to this section), each should be on its own line for consistency within the block.

File: src/AndersonAcceleration.f90 (hunk 16 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:343]: No, not for this guy. If depth <= 1 the linear solver is never inited because it isn't needed since the alphas can be calculated with a simple formula.

Comment (aarograh) [reply src/AndersonAcceleration.f90:343]: Oh, gotcha.  I missed the `%LS` part.  Thought that was the Anderson object you were checking.

File: src/AndersonAcceleration.f90 (hunk 15 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:324]: It will some day when Trilinos finished building.

File: src/AndersonAcceleration.f90 (hunk 14 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90]: So I'm fine with what you did on this... if it works.  If we were to build without trilinos and without mpi, though, it would compile to this:
```
SELECT TYPE(x)
ENDSELECT
```
Can you have a SELECT block with no branches?  I'm not even sure.  I'm fine leaving it this way as long as it works for a non-mpi, non-trilinos build.

File: src/AndersonAcceleration.f90 (hunk 15 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:322]: So I'm fine with what you did on this... if it works. If we were to build without trilinos and without mpi, though, it would compile to this:

SELECT TYPE(x)
ENDSELECT
Can you have a SELECT block with no branches? I'm not even sure. I'm fine leaving it this way as long as it works for a non-mpi, non-trilinos build.

File: src/AndersonAcceleration.f90 (hunk 10 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:100]: Also, is this something that would even be set up in XMLProc?  I feel like at most someone would simply enable anderson acceleration for a particular application, and then the code would have the appropriate values for that application.

Comment (jpjones6) [reply src/AndersonAcceleration.f90:100]: Any change in the problem (different geometry, power, flow, anything) will lead to a change in what's optimal for depth and beta (especially beta). The default for beta should be 0.5 because that is safe (beta acts as a damping factor), but it is likely far from optimal for most problems. So, really the user should be aware of what Anderson is set to if they are going to use it.

File: src/AndersonAcceleration.f90 (hunk 13 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:297]: We must be looking at different parts of the code entirely. I almost can't find an example of this in Futility or MPACT.

File: src/AndersonAcceleration.f90 (hunk 15 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:322]: Apparently yes. Worked locally. Good catch though looks janky.

File: src/AndersonAcceleration.f90 (hunk 10 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:100]: ok, that's fair.  But, for any of these values that could have a default, let's set a safe one here and then we can expose those to the user as needed (instead of ALWAYS requiring them to be set).  `beta` sounds like it could just default to 0.5.  It seems like `start` and `depth` could have some reasonable defaults too?  Obviously `n` has to be passed in, so that one is fine to REQUIRE the presence of.

File: src/AndersonAcceleration.f90 (hunk 13 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:297]: Well, that's how the standard is written.  Much of MPACT follows this.  Futility is probably much more a mixed bag since it's so old.  But yes, the rule is that the the dummy arguments each be on their own line immediately following the function/subroutine declaration line.  Other variables come after.

File: src/AndersonAcceleration.f90 (hunk 15 above)
Comment (aarograh) [reply src/AndersonAcceleration.f90:322]: Ok, cool.  If it works, it works.

File: src/AndersonAcceleration.f90 (hunk 10 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:100]: I'm good with that. I'll remove the defaults from XML_Proc on the MPACT side, but still warn for missing parameters there so the warnings are printed in the input processing. Does that sound good? Start can always be 1 (only time higher should really be chosen is when you know your initial guess is either garbage or not actually used for the vector you're accelerating such as when the initial vector is just defaulted to something). For depth 1 is safest, higher values for depth result in more aggressive acceleration that may or may not be stable or even faster.

Comment (aarograh) [reply src/AndersonAcceleration.f90:100]: Yeah that sounds good.  Thanks

File: src/AndersonAcceleration.f90 (hunk 13 above)
Comment (jpjones6) [reply src/AndersonAcceleration.f90:297]: Okay, taken care of.
